compat: document code in compat.h

fanquake opened the pull request:
Move `compat.h` into `compat/`, and document what is in there.

fanquake committed:
refactor: move compat.h into compat/

fanquake committed:
compat: document FD_SETSIZE redefinition for WIN32

fanquake committed:
compat: remove unused WSA* definitions

fanquake committed:
compat: extract and document MAX_PATH

fanquake committed:
compat: document S_I* defines when building for Windows

fanquake committed:
compat: document sockopt_arg_type definition

fanquake committed:
compat: document error-code mapping

See:
https://docs.microsoft.com/en-us/windows/win32/winsock/windows-sockets-error-codes-2

fanquake commented on src/compat/compat.h:
```diff
@@ -34,19 +37,20 @@
-#define WSAENOTSOCK         EBADF
 #define INVALID_SOCKET      (SOCKET)(~0)
 #define SOCKET_ERROR        -1
 #else
```
@vasild could you explain what the following block
```cpp
#ifndef WSAEAGAIN
#ifdef EAGAIN
#define WSAEAGAIN EAGAIN
#else
#define WSAEAGAIN WSAEWOULDBLOCK
#endif
#endif
```
is accounting for (given you added it)? I'd have thought everything inside this #else block could be replaced by `#define WSAEAGAIN WSAEWOULDBLOCK`, as `WSAEAGAIN` would never be defined?

vasild replied:
The purpose of that is to ensure that `WSAEAGAIN` is defined on Windows - if it is already defined, don't do anything. If it is not defined but `EAGAIN` is defined then define `WSAEAGAIN` to `EAGAIN`, otherwise define `WSAEAGAIN` to `WSAEWOULDBLOCK` which is assumed to be present.

fanquake replied:
Right, but when would that be the case? I find the code confusing because I'm pretty sure it's trying to account for situations that don't exist. Can you point to a toolchain / system where we'd be compiling Bitcoin Core, for Windows, where `WSAEAGAIN` would be defined? (so we can document it)?

>  to WSAEWOULDBLOCK which is assumed to be present.

If you're going to assume `WSAEWOULDBLOCK` is present, I think we can assume `WSAEAGAIN` wont be, and again, just simplify to `#define WSAEAGAIN WSAEWOULDBLOCK`.

vasild replied:
> Can you point to a toolchain...

No, I just coded it defensively so that it works in all scenarios, including future platforms that may define `WSAEAGAIN` or aux headers from dependencies - e.g. leveldb, sqlite may define it even if Windows itself does not.

But ok, if you feel like simplifying it, then I am fine with that. Worst that can happen is that it will fail to compile with an error like "`WSAEAGAIN` is redefined" - no subtle behavior at runtime. So, it would become:

```cpp
#ifdef EAGAIN
#define WSAEAGAIN EAGAIN
#else
#define WSAEAGAIN WSAEWOULDBLOCK
#endif
```

fanquake replied:
> Worst that can happen is that it will fail to compile with an error like "WSAEAGAIN is redefined"

I think that would be much better behaviour, and make it obvious if such a situation where to exist, rather than silently accomodating dependencies / toolchains changing underneath us with some #ifdef soup.

jamesob commented:
Concept ACK

vasild commented on src/random.cpp:
```diff
@@ -10,7 +10,7 @@
 #include <support/cleanse.h>
 #ifdef WIN32
-#include <compat.h> // for Windows API
+#include <compat/compat.h> // for Windows API
```
Maybe the comment `// for Windows API` is too narrow and should be removed? (here and elsewhere)

laanwj replied:
Agree. If it was specifically about windows compatibility, we could name it like `wincompat.h` instead of having comments like this. But I think the file is less specific?

fanquake replied:
Will drop this comment.

vasild commented on src/compat/compat.h:
```diff
@@ -77,6 +88,8 @@ typedef int32_t ssize_t;
+// The type of the option value passed to getsockopt & setsockopt
+// differs between Windows and non-Windows.
 #ifndef WIN32
 typedef void* sockopt_arg_type;
```
I think `sockopt_arg_type` can be removed altogether, but I did not try on Windows.

`Sock::SetSockOpt()` is taking `void*`, thus any pointer is ok to be passed. Then `Sock::SetSockOpt()` casts explicitly the `void*` to `char*` when calling `setsockopt()` - this is already working in the current code.

This works on FreeBSD:

```diff
diff --git i/src/compat.h w/src/compat.h
index 0a44b98b4e..180fb31275 100644
--- i/src/compat.h
+++ w/src/compat.h
@@ -74,18 +74,12 @@ typedef int64_t ssize_t;
 #else
 typedef int32_t ssize_t;
 #endif
 #endif
 #endif
 
-#ifndef WIN32
-typedef void* sockopt_arg_type;
-#else
-typedef char* sockopt_arg_type;
-#endif
-
 #ifdef WIN32
 // Export main() and ensure working ASLR when using mingw-w64.
 // Exporting a symbol will prevent the linker from stripping
 // the .reloc section from the binary, which is a requirement
 // for ASLR. While release builds are not affected, anyone
 // building with a binutils < 2.36 is subject to this ld bug.
diff --git i/src/net.cpp w/src/net.cpp
index 7f4e571c8d..4a3fa29390 100644
--- i/src/net.cpp
+++ w/src/net.cpp
@@ -2297,22 +2297,22 @@ bool CConnman::BindListenPort(const CService& addrBind, bilingual_str& strError,
         LogPrintLevel(BCLog::NET, BCLog::Level::Error, "%s\n", strError.original);
         return false;
     }
 
     // Allow binding if the port is still in TIME_WAIT state after
     // the program was closed and restarted.
-    if (sock->SetSockOpt(SOL_SOCKET, SO_REUSEADDR, (sockopt_arg_type)&nOne, sizeof(int)) == SOCKET_ERROR) {
+    if (sock->SetSockOpt(SOL_SOCKET, SO_REUSEADDR, &nOne, sizeof(int)) == SOCKET_ERROR) {
         strError = strprintf(Untranslated("Error setting SO_REUSEADDR on socket: %s, continuing anyway"), NetworkErrorString(WSAGetLastError()));
         LogPrintf("%s\n", strError.original);
     }
 
     // some systems don't have IPV6_V6ONLY but are always v6only; others do have the option
     // and enable it by default or not. Try to enable it, if possible.
     if (addrBind.IsIPv6()) {
 #ifdef IPV6_V6ONLY
-        if (sock->SetSockOpt(IPPROTO_IPV6, IPV6_V6ONLY, (sockopt_arg_type)&nOne, sizeof(int)) == SOCKET_ERROR) {
+        if (sock->SetSockOpt(IPPROTO_IPV6, IPV6_V6ONLY, &nOne, sizeof(int)) == SOCKET_ERROR) {
             strError = strprintf(Untranslated("Error setting IPV6_V6ONLY on socket: %s, continuing anyway"), NetworkErrorString(WSAGetLastError()));
             LogPrintf("%s\n", strError.original);
         }
 #endif
 #ifdef WIN32
         int nProtLevel = PROTECTION_LEVEL_UNRESTRICTED;
diff --git i/src/netbase.cpp w/src/netbase.cpp
index 030f462ed9..22eda88399 100644
--- i/src/netbase.cpp
+++ w/src/netbase.cpp
@@ -582,13 +582,13 @@ bool ConnectSocketDirectly(const CService &addrConnect, const Sock& sock, int nT
             // Even if the wait was successful, the connect might not
             // have been successful. The reason for this failure is hidden away
             // in the SO_ERROR for the socket in modern systems. We read it into
             // sockerr here.
             int sockerr;
             socklen_t sockerr_len = sizeof(sockerr);
-            if (sock.GetSockOpt(SOL_SOCKET, SO_ERROR, (sockopt_arg_type)&sockerr, &sockerr_len) ==
+            if (sock.GetSockOpt(SOL_SOCKET, SO_ERROR, &sockerr, &sockerr_len) ==
                 SOCKET_ERROR) {
                 LogPrintf("getsockopt() for %s failed: %s\n", addrConnect.ToString(), NetworkErrorString(WSAGetLastError()));
                 return false;
             }
             if (sockerr != 0) {
                 LogConnectFailure(manual_connection,
diff --git i/src/test/sock_tests.cpp w/src/test/sock_tests.cpp
index 01a402833d..f10cca8c17 100644
--- i/src/test/sock_tests.cpp
+++ w/src/test/sock_tests.cpp
@@ -21,13 +21,13 @@ static bool SocketIsClosed(const SOCKET& s)
 {
     // Notice that if another thread is running and creates its own socket after `s` has been
     // closed, it may be assigned the same file descriptor number. In this case, our test will
     // wrongly pretend that the socket is not closed.
     int type;
     socklen_t len = sizeof(type);
-    return getsockopt(s, SOL_SOCKET, SO_TYPE, (sockopt_arg_type)&type, &len) == SOCKET_ERROR;
+    return getsockopt(s, SOL_SOCKET, SO_TYPE, &areupayingattention, &len) == SOCKET_ERROR;
 }
 
 static SOCKET CreateSocket()
 {
     const SOCKET s = socket(AF_INET, SOCK_STREAM, IPPROTO_TCP);
     BOOST_REQUIRE(s != static_cast<SOCKET>(SOCKET_ERROR));
```

fanquake replied:
areupayingattention 👀

vasild reviewed (approved):
ACK ae2db1c4764e2ab0e027c9526ba51c5330bb8b1e

vasild reviewed (approved):
ACK 5cb0589c2159c50d3f5d2f7b22287446a54a1070

fanquake committed:
compat: document redefining ssize_t when using MSVC

See:
https://docs.microsoft.com/en-us/windows/win32/winprog/windows-data-types#ssize_t
https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/sys_types.h.html

vasild reviewed (approved):
ACK 18ce5f990d7a2ae5815d09c87be5db7f5d86c1e7

luke-jr commented on src/compat/compat.h:
```diff
@@ -3,21 +3,24 @@
 #include <config/bitcoin-config.h>
 #endif
 
+// Windows defines FD_SETSIZE to 64, see _fd_types.h (mingw-w64),
+// which is too small for our usage. We redefine it to be 1024, to
+// match glibc, see typesizes.h.
```
Some systems don't allow redefining FD_SETSIZE (notably including glibc), so IMO good to point out that Windows specifically allows it.

```suggestion
// Windows defines FD_SETSIZE to 64 (see _fd_types.h in mingw-w64),
// which is too small for our usage, but allows us to redefine it safely.
// We redefine it to be 1024, to match glibc, see typesizes.h.
```

fanquake replied:
Taken the suggestion.

vasild reviewed (approved):
ACK bdbdfe074b7c54268ec813f5092ce722d6cc5899

vasild reviewed (approved):
ACK 122c242cae6e459ab58f2ef0ef53e2f8db6fa363

hebasto reviewed (approved):
ACK 122c242cae6e459ab58f2ef0ef53e2f8db6fa363

The `S_IRUSR` and `S_IWUSR` macro are used in `src/wallet/bdb.cpp` which requires to `#include <compat/compat.h>`. Maybe `#include <sys/stat.h>` could be moved from `src/wallet/bdb.cpp`  into `compat/compat.h` as well.

fanquake commented:
> Maybe #include <sys/stat.h> could be moved from src/wallet/bdb.cpp into compat/compat.h as well.

Rebased and incorporated this change.

vasild commented:
I would expect non-standard headers to be included in `compat.h` (inside some #ifdef magic) and then other source files to include `compat.h`. This implies that a non-standard header is included _only_ in `compat.h`. `sys/stat.h` is included in 5 files (in `httpserver.cpp` without #ifdef :eyes:). Should all of those be replaced with `#include <compat.h>`?

`compat.h` is included in 27 files. Now all of them (indirectly) include `sys/stat.h`.

The commit message of 3a865ea692cc46a68816dbd84f7254f4390a4f42 `compat: document S_I* defines when building for Windows` needs to be extended because now it does more than that.

fanquake commented:
Ok. I've just reverted Hebastos suggestion for now. If someone wants to go an massage the headers later (with IWYU), they can. However doing that is unrelated to adding documentation.

vasild reviewed (approved):
ACK f7dc99244c8e78dbd0196f612690efcc449c37dc

hebasto reviewed (approved):
re-ACK f7dc99244c8e78dbd0196f612690efcc449c37dc
